libs/ts-rest/core/src/lib/type-utils.ts

```diff
@@ -103,3 +106,28 @@ export type Prettify<T> = {
   [K in keyof T]: T[K];
   // eslint-disable-next-line @typescript-eslint/ban-types
 } & {};
+
+export type DefinedOrEmpty<
+  T,
+  K extends keyof NonNullable<T>
+  // eslint-disable-next-line @typescript-eslint/ban-types
+> = undefined extends T ? {} : NonNullable<T>[K];
+
+declare const tag: unique symbol;
+
+declare type Tagged<Token> = {
+  readonly [tag]: Token;
+};
+
+export type Opaque<Type, Token = unknown> = Type & Tagged<Token>;
+
+export type WithoutUnknown<T> = Pick<
+  T,
+  {
+    [K in keyof T]: unknown extends Exclude<T[K], undefined> ? never : K;
+  }[keyof T]
+>;
+
+export type LowercaseKeys<T> = Prettify<{
```

Comment:
Since we're doing Lowercase type casting for header keys, where are we ensuring that the header keys are lowercase? Is it after you initialize the `router` or during the consumption of the header (in say, an express server endpoint for instance)?

Reply:
Express `req.headers` is always normalized in lowercase so that's why we need to ensure the accessor types are in lowercase too on the server-side. On the client side it wouldn't really matter if we weren't merging types, but since we do, if we had the same header with different casing in `baseHeaders` and an endpoint's `headers` they would be included together as two separate properties.